kaldb/src/main/java/com/slack/kaldb/logstore/search/LogIndexSearcherImpl.java

```diff
@@ -111,30 +129,39 @@ public SearchResult<LogMessage> search(
       // This is a useful optimization for indexes that are static.
       IndexSearcher searcher = searcherManager.acquire();
       try {
-        TopFieldCollector topFieldCollector = buildTopFieldCollector(howMany);
-        StatsCollector statsCollector =
+        CollectorManager<TopFieldCollector, TopFieldDocs> topFieldCollector =
+            buildTopFieldCollector(howMany);
+        CollectorManager<StatsCollector, Histogram> statsCollector =
             buildStatsCollector(bucketCount, startTimeMsEpoch, endTimeMsEpoch);
-        Collector collectorChain = MultiCollector.wrap(topFieldCollector, statsCollector);
 
-        searcher.search(query, collectorChain);
+        MultiCollectorManager collectorManager = new MultiCollectorManager(topFieldCollector);
+        if (bucketCount > 0) {
+          collectorManager = new MultiCollectorManager(topFieldCollector, statsCollector);
+        }
+
         List<LogMessage> results;
+        Histogram histogram = new NoOpHistogramImpl();
         if (howMany > 0) {
-          ScoreDoc[] hits = topFieldCollector.topDocs().scoreDocs;
+          Object[] collector = searcher.search(query, collectorManager);
+
+          ScoreDoc[] hits = ((TopFieldDocs) collector[0]).scoreDocs;
           results = new ArrayList<>(hits.length);
           for (ScoreDoc hit : hits) {
             results.add(buildLogMessage(searcher, hit));
           }
+          if (bucketCount > 0) {
+            histogram = ((Histogram) collector[1]);
+          }
         } else {
           results = Collections.emptyList();
+          histogram = searcher.search(query, statsCollector);
         }
 
-        Histogram histogram = statsCollector.histogram;
-
         elapsedTime.stop();
         return new SearchResult<>(
             results,
             elapsedTime.elapsed(TimeUnit.MICROSECONDS),
-            histogram.count(),
+            bucketCount > 0 ? histogram.count() : results.size(),
```

Comment:
niceee!! This is likely what will cause the perf improvements since lucene can short circuit